### PR TITLE
[#1497] improvement(client): flushing buffer if the memoryUsed of the first record of `WriterBuffer` larger than bufferSize

### LINKED ISSUE #1497: [Improvement] flushing buffer if the memoryUsed of the first record of WriterBuffer larger than bufferSize
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-uniffle/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

Currently in `org.apache.spark.shuffle.writer.WriteBufferManager#insertIntoBuffer`, if the memoryUsed of WriterBuffer larger than bufferSize, the WriterBuffer will be flush. But if the memoryUsed of the first record larger than bufferSize, the WriterBuffer will not be flush.

### How should we improve?

Flushing buffer if the memoryUsed of the first record of WriterBuffer larger than bufferSize

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by xianjingfeng**
### What changes were proposed in this pull request?

Flushing buffer if the memoryUsed of the first record of `WriterBuffer` larger than bufferSize.

### Why are the changes needed?

More accurate. fix #1497 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

UT.

### CODE REVIEW

**Comment by zuston** [comment]
`hasRequested` is needed, please see the original comment.

**Comment by xianjingfeng** [comment]
> `hasRequested` is needed, please see the original comment.

Updated.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java [hunk 1]
```diff
@@ -201,39 +202,21 @@ private List<ShuffleBlockInfo> insertIntoBuffer(
...
+    if (wb != null) {
       if (wb.askForMemory(serializedDataLength)) {
         requestMemory(required);
         hasRequested = true;
       }
     }
 
-    if (buffers.containsKey(partitionId)) {
+    // The clear operation which triggered by spill operation will cause thread-safe problems
+    wb = buffers.remove(partitionId);
```
Comment (xianjingfeng) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:214]: There seems to be something wrong with the original logic. Such as the following scene.

`buffers.get()[thread-clear] ->  buffers.get()[thread-insert] -> 
wb.getData()[thread-clear] -> wb.addRecord()[thread-insert] -> wb.getMemoryUsed()[thread-clear] (maybe cause free more memory and miss record)`

`buffers.get()[thread-clear] ->  buffers.get()[thread-insert] -> 
wb.getData()[thread-clear] -> wb.getMemoryUsed()[thread-clear] -> wb.addRecord()[thread-insert] (maybe cause miss record)`

PTAL @zuston @jerqi

Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:214]: Could you help add the test case to verify this? I'm trying to understand this.

Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:214]: The test case is successful in current master branch, so I'm still not catch your thought about this patch.

Comment (xianjingfeng) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:214]: > The test case is successful in current master branch, so I'm still not catch your thought about this patch.

Please repeat this test for 100 times in current master branch. @zuston

Comment (xianjingfeng) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:214]: Snipaste_2024-01-30_16-17-14

Comment (xianjingfeng) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java]: It is my fault.

**Review by zuston — COMMENTED** [review]
Overall LGTM. Please add tests to cover this case, and also please create a issue to link this, that I don't think this is a minor fix.

**Review by zuston — APPROVED** [review]
LGTM. Thanks @xianjingfeng for your contribution!

**Pull request merged**